src/node_embedding_api.h

```diff
@@ -0,0 +1,398 @@
+    napi_value process,
+    napi_value require);
+
+typedef napi_value(NAPI_CDECL* node_embedding_initialize_module_callback)(
+    node_embedding_runtime runtime,
+    void* cb_data,
+    napi_env env,
+    const char* module_name,
+    napi_value exports);
+
+typedef void(NAPI_CDECL* node_embedding_event_loop_handler)(
+    node_embedding_runtime runtime, void* handler_data);
+
+typedef void(NAPI_CDECL* node_embedding_node_api_callback)(
+    node_embedding_runtime runtime, void* cb_data, napi_env env);
+
+//==============================================================================
+// Functions
+//==============================================================================
+
+//------------------------------------------------------------------------------
+// Node.js main function.
+//------------------------------------------------------------------------------
+
+// Runs Node.js main function as if it is invoked from Node.js CLI without any
+// embedder customizations.
+NAPI_EXTERN node_embedding_exit_code NAPI_CDECL
+node_embedding_run_nodejs_main(int32_t argc, char* argv[]);
+
+//------------------------------------------------------------------------------
+// Error handling functions.
+//------------------------------------------------------------------------------
+
+// Sets the global error handing for the Node.js embedding API.
+NAPI_EXTERN node_embedding_exit_code NAPI_CDECL node_embedding_on_error(
+    node_embedding_error_handler error_handler, void* error_handler_data);
+
+//------------------------------------------------------------------------------
+// Node.js global platform functions.
+//------------------------------------------------------------------------------
+
+// Creates a new Node.js platform instance.
+NAPI_EXTERN node_embedding_exit_code NAPI_CDECL node_embedding_create_platform(
+    int32_t api_version, node_embedding_platform* result);
+
+// Deletes the Node.js platform instance.
+NAPI_EXTERN node_embedding_exit_code NAPI_CDECL
+node_embedding_delete_platform(node_embedding_platform platform);
+
+// Checks if the Node.js platform is initialized.
+NAPI_EXTERN node_embedding_exit_code NAPI_CDECL
+node_embedding_platform_is_initialized(node_embedding_platform platform,
+                                       bool* result);
+
+// Sets the flags for the Node.js platform initialization.
+NAPI_EXTERN node_embedding_exit_code NAPI_CDECL
+node_embedding_platform_set_flags(node_embedding_platform platform,
+                                  node_embedding_platform_flags flags);
+
+// Sets the CLI arguments for the Node.js platform initialization.
+NAPI_EXTERN node_embedding_exit_code NAPI_CDECL
+node_embedding_platform_set_args(node_embedding_platform platform,
+                                 int32_t argc,
+                                 char* argv[]);
+
+// Initializes the Node.js platform.
+NAPI_EXTERN node_embedding_exit_code NAPI_CDECL
+node_embedding_platform_initialize(node_embedding_platform platform,
```

Comment:
Flags and args are unsafe to be changed after initialization. I think `node_embedding_platform_set_flags` and `node_embedding_platform_set_args` should be merged in `node_embedding_platform_initialize`.

Reply:
Good point! I was also thinking along the same lines. As a result the "initialization" and "is_initialized" methods are removed and the "create" methods have a new configuration callback where all the builder set methods are called.
Based on your feedback I am going to go one more step further and introduce opaque "options" types so that all these "set_" and "on_" methods are only applicable to these "optons" types.
This way it would be easy to understand when it is safe to call them.